zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/QuorumZooKeeperServer.java

```diff
@@ -189,9 +189,18 @@ public void dumpConf(PrintWriter pwriter) {
         pwriter.print(self.getQuorumVerifier().toString());
     }
 
+    private final Object stateChangeMutex = new Object();
+
     @Override
     protected void setState(State state) {
```

Comment:
After a closer look I find that setState can be a single implementation in `ZooKeeperServer`. Just as is there. And you can apply this check as well as logging to verify as @eolivelli said.

Reply:
> But @Tison's point still holds; it seems to me that this could be simplified by making `setState()` a `synchronized` method. Do you have a specific reason of using a separate object?
> 
> (The parent `setState()` does not _need_ to be `synchronized` because the field is `volatile`.)

@ztzg The rationale for not making `setState()` a synchronized method is that if a thread is running another synchronized method of this ZooKeeperServer object for some time, it may block the `setState()` invoked by another thread, and this invocation may be critical.

Reply:
> I agree with @eolivelli that `setState()` has to be "careful," because `ZooKeeperServerListenerImpl.notifyStopping` calls it from "random" threads.

Actually, the single-node ZooKeeperServer can handle this issue with the handler:
https://github.com/apache/zookeeper/blob/0c98d1d3252e645a5c25bfecaba5ca1f5cd3258e/zookeeper-server/src/main/java/org/apache/zookeeper/server/ZooKeeperServer.java#L779-L789
because it is registered at the very beginning:
https://github.com/apache/zookeeper/blob/0c98d1d3252e645a5c25bfecaba5ca1f5cd3258e/zookeeper-server/src/main/java/org/apache/zookeeper/server/ZooKeeperServerMain.java#L150
This handling logic is:
https://github.com/apache/zookeeper/blob/0c98d1d3252e645a5c25bfecaba5ca1f5cd3258e/zookeeper-server/src/main/java/org/apache/zookeeper/server/ZooKeeperServerShutdownHandler.java#L42-L46
and then
https://github.com/apache/zookeeper/blob/0c98d1d3252e645a5c25bfecaba5ca1f5cd3258e/zookeeper-server/src/main/java/org/apache/zookeeper/server/ZooKeeperServerMain.java#L181-L185

However, in the case of `QuorumZooKeeperServer`, including Leader, Follower, etc., the aforementioned logic is not used.